applications/ShapeOptimizationApplication/python_scripts/mapping/sliding_vertex_morphing_mapper.py

```diff
@@ -0,0 +1,393 @@
+                                "mapper_type": "nearest_element",
+                                "echo_level" : 0
+                            }""")
+        main_mapper = SpacialMapperFactory.CreateMapper(
+            structural_model.GetModelPart("main"), self.destination_model_part,
+            ne_mapper_settings)
+        main_mapper.Map(KM.DISPLACEMENT, input_variable)
+
+    def _CreateStructuralModel(self, structural_model):
+
+        main_part = structural_model.CreateModelPart("main")
+        mesh_motion_conditions = main_part.CreateSubModelPart("prescribed_bc")
+        fixed_nodes = main_part.CreateSubModelPart("fixed_bc")
+
+        # add solution variables necessary for structural mechanics analysis
+        main_part.AddNodalSolutionStepVariable(KSO.SHAPE_UPDATE)
+        main_part.AddNodalSolutionStepVariable(KM.DISPLACEMENT)
+        main_part.AddNodalSolutionStepVariable(KM.ROTATION)
+        main_part.AddNodalSolutionStepVariable(KM.REACTION)
+        main_part.AddNodalSolutionStepVariable(KM.POSITIVE_FACE_PRESSURE)
+        main_part.AddNodalSolutionStepVariable(KM.NEGATIVE_FACE_PRESSURE)
+        main_part.AddNodalSolutionStepVariable(KSM.POINT_LOAD)
+        main_part.AddNodalSolutionStepVariable(KSM.LINE_LOAD)
+        main_part.AddNodalSolutionStepVariable(KSM.SURFACE_LOAD)
+        main_part.AddNodalSolutionStepVariable(KM.VOLUME_ACCELERATION)
+        main_part.AddNodalSolutionStepVariable(KM.ROTATION)
+        main_part.AddNodalSolutionStepVariable(KM.REACTION_MOMENT)
+        main_part.AddNodalSolutionStepVariable(KSM.POINT_MOMENT)
+        main_part.AddNodalSolutionStepVariable(KSO.NORMALIZED_SURFACE_NORMAL)
+
+        material_properties = main_part.GetProperties()[0]
+        self._ApplyMaterialProperties(material_properties)
+
+        # create nodes of prescribed bc
+        for node in self.sliding_mesh.Nodes:
+            mesh_motion_conditions.CreateNewNode(node.Id, node.X, node.Y, node.Z)
+
+        # Identify mesh motion area / nodes to be damped
+        KM.VariableUtils().SetFlag(KM.STRUCTURE, False, main_part.Nodes)
+        radius = self.settings["filter_radius"].GetDouble()
+        search_based_functions = KSO.SearchBasedFunctions(self.destination_model_part)
+        search_based_functions.FlagNodesInRadius(mesh_motion_conditions.Nodes, KM.STRUCTURE, 2*radius)
+
+        # create nodes for main_part
+        for node in self.destination_model_part.Nodes:
+            if node.Is(KM.STRUCTURE):
+                main_part.CreateNewNode(node.Id, node.X, node.Y, node.Z)
+        # create elements for main_part
+        ele_id = 0
+        node_ids_to_remove = []
+        for surface_condition in self.destination_model_part.Conditions:
+            node_counter = 0
+            number_of_nodes = len(surface_condition.GetNodes())
+            ele_node_ids = []
+            for node in surface_condition.GetNodes():
+                if main_part.HasNode(node.Id):
```

Comment:
please store the node_ids that you have previously added in a set and than look it up there, that should be more efficient.